pulp/apis/xpress_api.py

```diff
@@ -318,3 +325,424 @@ def quote_path(path):
+                    7: constants.LpStatusUndefined,  # XPRS_MIP_UNBOUNDED
+                }
+                statuskey = "mipstatus"
+            else:
+                # Solved as continuous
+                x, slacks, duals, djs = [], [], [], []
+                try:
+                    model.getlpsol(x, slacks, duals, djs)
+                except:
+                    # No solution available
+                    x, slacks, duals, djs = None, None, None, None
+                statusmap = {
+                    0: constants.LpStatusNotSolved,  # XPRS_LP_UNSTARTED
+                    1: constants.LpStatusOptimal,  # XPRS_LP_OPTIMAL
+                    2: constants.LpStatusInfeasible,  # XPRS_LP_INFEAS
+                    3: constants.LpStatusUndefined,  # XPRS_LP_CUTOFF
+                    4: constants.LpStatusUndefined,  # XPRS_LP_UNFINISHED
+                    5: constants.LpStatusUnbounded,  # XPRS_LP_UNBOUNDED
+                    6: constants.LpStatusUndefined,  # XPRS_LP_CUTOFF_IN_DUAL
+                    7: constants.LpStatusNotSolved,  # XPRS_LP_UNSOLVED
+                    8: constants.LpStatusUndefined,  # XPRS_LP_NONCONVEX
+                }
+                statuskey = "lpstatus"
+            if x is not None:
+                lp.assignVarsVals({v.name: x[v._xprs[0]] for v in lp.variables()})
+            if djs is not None:
+                lp.assignVarsDj({v.name: djs[v._xprs[0]] for v in lp.variables()})
+            if duals is not None:
+                lp.assignConsPi(
+                    {c.name: duals[c._xprs[0]] for c in lp.constraints.values()}
+                )
+            if slacks is not None:
+                lp.assignConsSlack(
+                    {c.name: slacks[c._xprs[0]] for c in lp.constraints.values()}
+                )
+
+            status = statusmap.get(
+                model.getAttrib(statuskey), constants.LpStatusUndefined
+            )
+            lp.assignStatus(status)
+
+            # Mark all variables and constraints as unmodified so that
+            # actualResolve will do the correct thing.
+            for v in lp.variables():
+                v.modified = False
+            for c in lp.constraints.values():
+                c.modified = False
+
+            return status
+
+        except (xpress.ModelError, xpress.InterfaceError, xpress.SolverError) as err:
+            raise PulpSolverError(str(err))
+
+    def actualSolve(self, lp, prepare=None):
+        """Solve a well formulated lp problem"""
+        if not self.available():
+            # Import again to get a more verbose error message
+            message = "XPRESS Python API not available"
+            try:
+                import xpress
+            except ImportError as err:
+                message = str(err)
+            raise PulpSolverError(message)
+
+        self._extract(lp)
+        try:
+            model = lp._xprs
+            # Apply controls that were passed to the constructor
```

Comment:
it would be nice to keep the `GUROBI`, `CPLEX_PY` convention of using `self.solverModel` to store the solver object. Also, I'd prefer to have the same convention of having the `actualSolve` that does a "simple solve" and then having the possibility (as documented [here](https://coin-or.github.io/pulp/guides/how_to_configure_solvers.html#using-solver-specific-functionality)) to split the solving into steps in order to tweak the solver object. The steps should be: (1) build the solver model, (2) tweak it, (3) call the solver.

Reply:
I changed the code as requested.
- Renamed `lp._xprs` to `lp.solverModel`
- Split the solution process into `buildSolverModel`, `callSolver`, `findSolutionValues`